tpmutil/mssim/mssim.go

```diff
@@ -0,0 +1,227 @@
+// Package mssim implements the Microsoft simulator TPM2 Transmission Interface
+//
+// The Microsoft simulator TPM Command Transmission Interface (TCTI) is a
+// remote procedure interface donated to the TPM2 Specification by Microsoft.
+// Its primary implementation is the tpm_server maintained by IBM.
+//
+// https://sourceforge.net/projects/ibmswtpm2/
+//
+// This package implements client code to communicate with server code described
+// in the document "TPM2 Specification Part 4: Supporting Routines – Code"
+//
+// https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-4-Supporting-Routines-01.38-code.pdf
+package mssim
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync/atomic"
+)
+
+// Constants defined in "D.3.2. Typedefs and Defines"
+const (
+	tpmSignalPowerOn  = 1
+	tpmSignalPowerOff = 2
+	tpmSendCommand    = 8
+	tpmSignalNVOn     = 11
+	tpmSessionEnd     = 20
+)
+
+// Config holds configuration parameters for connecting to the simluator.
+type Config struct {
+	// Addresses of the command and platform handlers.
+	//
+	// Defaults to port 2321 and 2322.
+	CommandAddress  string
+	PlatformAddress string
+
+	// DialContext is used to connect to the simluator ports.
+	//
+	// If nil, a reasonable default is provided.
+	DialContext func(ctx context.Context, network, address string) (net.Conn, error)
+}
+
+// Open attempts to initialize the simulator, power cycling the platform.
+func Open(ctx context.Context, config Config) (*Conn, error) {
+	cmdAddr := config.CommandAddress
+	if cmdAddr == "" {
+		cmdAddr = "127.0.0.1:2321"
+	}
+
+	platformAddr := config.PlatformAddress
+	if platformAddr == "" {
+		platformAddr = "127.0.0.1:2322"
+	}
+
+	dial := config.DialContext
+	if dial == nil {
+		dial = (&net.Dialer{}).DialContext
+	}
+
+	conn, err := dial(ctx, "tcp", platformAddr)
+	if err != nil {
+		return nil, fmt.Errorf("dial platform address: %v", err)
+	}
+	defer conn.Close()
+
+	// Startup the simulator.
```

Comment:
Can you explain the startup procedure in a comment?
It's not clear to me why it has to be power cycled and what sessionEnd does

Reply:
Added a comment. Basically, the only poweron command I could find that worked was part of some IBM TPM2 utilities. https://sourceforge.net/projects/ibmtpm20tss/

Based off wireshark, those are the commands it's executing.

It's been fun dealing with poorly documented c programs hosted on source forge...